src/Tickets/Emails/Dispatcher.php

```diff
@@ -0,0 +1,507 @@
+		return $this->email;
+	}
+
+	/**
+	 * Prepares the current dispatcher for sending an email.
+	 *
+	 * @since TBD
+	 *
+	 * @param Email_Abstract $email Email Instance used to prepare the dispatcher with.
+	 */
+	protected function prepare_dispatcher( Email_Abstract $email ): void {
+		// Enforce text/html content type header.
+		$this->add_header( 'Content-Type', 'text/html; charset=utf-8' );
+		$this->set_email( $email );
+
+		$from_email = $email->get_from_email();
+		$from_name  = $email->get_from_name();
+
+		// Add From name/email to headers if no headers set yet, and we have a valid From email address.
+		if ( ! empty( $from_name ) && ! empty( $from_email ) && is_email( $from_email ) ) {
+			$from_email = sanitize_email( $from_email );
+
+			$this->add_header( 'From', sprintf(
+				'%1$s <%2$s>',
+				stripcslashes( $from_name ),
+				$from_email
+			) );
+
+			$this->add_header( 'Reply-To', $from_email );
+		}
+
+		$this->set_to( $email->get_recipient() );
+		$this->set_subject( $email->get_subject() );
+		$this->set_content( $email->get_content() );
+	}
+
+	/**
+	 * From a given email type instance generate a new Dispatcher and prepare it for usage.
+	 *
+	 * @since TBD
+	 *
+	 * @param Email_Abstract $email
+	 *
+	 * @return Dispatcher
+	 */
+	public static function from_email( Email_Abstract $email ): Dispatcher {
+		// Generate a new dispatcher every time.
+		$dispatcher = tribe( static::class );
+
+		// Prepare the dispatcher to send an email.
+		$dispatcher->prepare_dispatcher( $email );
+
+		/**
+		 * Allows modifications of the Email Dispatcher to all Email Types.
+		 *
+		 * @since TBD
+		 *
+		 * @param Dispatcher     $dispatcher Which dispatcher instance will be used for the email sent.
+		 * @param Email_Abstract $email      Which instance of the email that will be attached to this dispatcher.
+		 */
+		$dispatcher = apply_filters( 'tec_tickets_emails_dispatcher', $dispatcher, $email );
+
+		$email_slug = $email->slug;
+
+		/**
+		 * Allows modifications of the Email Dispatcher specific to this Email Type.
+		 *
+		 * @since TBD
+		 *
+		 * @param Dispatcher     $dispatcher Which dispatcher instance will be used for the email sent.
+		 * @param Email_Abstract $email      Which instance of the email that will be attached to this dispatcher.
+		 */
+		return apply_filters( "tec_tickets_emails_{$email_slug}_dispatcher", $dispatcher, $email );
+	}
+
+	/**
+	 * Includes a new header to be dispatched.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $key
+	 * @param string $value
+	 */
+	public function add_header( string $key, string $value ): void {
+		$this->headers[ $key ] = $value;
+	}
+
+	/**
+	 * Includes a new attachment to be dispatched.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $path
+	 */
+	public function add_attachment( string $path ): void {
+		$this->attachments[] = $path;
+	}
+
+	/**
+	 * Get dispatcher headers.
+	 *
+	 * @since TBD
+	 *
+	 * @return array
+	 */
+	public function get_headers(): array {
```

Comment:
Why don't you keep the headers method in the email class (where it belongs, because it's tied to emails) and grab the headers from there by using the `$this->get_email()->get_headers()`?

Naming this with `dispatcher` is adding more noise, and it makes it more difficult to have access to the actual email if you're trying (as a dev that's extending the functionality) to get access to the actual email.

Reply:
We are separating concerns, Email Type class is where currently we hold data, and eventually refactored to just hold the pattern of how to generate an email, not the actual data of the email or the sending of the email.